In [5]:
import pandas as pd
import numpy as np
from pydataset import data
from env import host, user, password

In [6]:
def get_db_url(host,user,password):
    datatable = input('Please enter the database you wish to connect to: ')
    url = f'mysql+pymysql://{user}:{password}@{host}/{datatable}'
    return url

In [7]:
url = get_db_url(host,user,password)
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

Please enter the database you wish to connect to: employees


emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

def get_db_url2(host,user,password):
    
    url = f'mysql+pym1ysql://{user}:{password}@{host}/employees'
    return url

In [8]:
url2 = get_db_url2(host,user,password)
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url2)  ##3 If we make a type we get a can't load plugin error

NameError: name 'get_db_url2' is not defined

In [9]:
url = get_db_url(host,user,password)
pd.read_sql('SELECCT * FROM employees LIMIT 5 OFFSET 50', url) ##intentional error for error message review

Please enter the database you wish to connect to: employees


ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'SELECCT * FROM employees LIMIT 5 OFFSET 50' at line 1")
[SQL: SELECCT * FROM employees LIMIT 5 OFFSET 50]
(Background on this error at: http://sqlalche.me/e/13/f405)

Question 4: Read the employees and titles tables into two separate DataFrames

In [10]:
employees_df = pd.read_sql('SELECT * FROM employees', url)
employees_df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [11]:
type(employees_df)

pandas.core.frame.DataFrame

In [12]:
titles_df = pd.read_sql('SELECT * FROM titles', url)
titles_df

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

Question 5: How many rows and columns do you have in each dataframe?

In [13]:
print(employees_df.shape)
print(titles_df.shape)

(300024, 6)
(443308, 4)


Question 6: Display the summary statistics for each dataframe?

In [14]:
employees_df.describe(include='all')

emp_no  birth_date first_name last_name  gender   hire_date
count   300024.000000      300024     300024    300024  300024      300024
unique            NaN        4750       1275      1637       2        5434
top               NaN  1952-03-08     Shahab      Baba       M  1985-06-20
freq              NaN          95        295       226  179973         132
mean    253321.763392         NaN        NaN       NaN     NaN         NaN
std     161828.235540         NaN        NaN       NaN     NaN         NaN
min      10001.000000         NaN        NaN       NaN     NaN         NaN
25%      85006.750000         NaN        NaN       NaN     NaN         NaN
50%     249987.500000         NaN        NaN       NaN     NaN         NaN
75%     424993.250000         NaN        NaN       NaN     NaN         NaN
max     499999.000000         NaN        NaN       NaN     NaN         NaN

In [15]:
titles_df.describe(include='all')

emp_no     title   from_date     to_date
count   443308.000000    443308      443308      443308
unique            NaN         7        6393        5888
top               NaN  Engineer  1998-10-25  9999-01-01
freq              NaN    115003         132      240124
mean    253075.034430       NaN         NaN         NaN
std     161853.292613       NaN         NaN         NaN
min      10001.000000       NaN         NaN         NaN
25%      84855.750000       NaN         NaN         NaN
50%     249847.500000       NaN         NaN         NaN
75%     424891.250000       NaN         NaN         NaN
max     499999.000000       NaN         NaN         NaN

Question 7: How many unique titles are in the titles DataFrame

In [16]:
len(titles_df['title'].value_counts())

7

Question 8: What is the oldest_date in the to_date column?

In [17]:
titles_df['to_date'].sort_values().head(1) # Oldest to_date date below

16064    1985-03-01
Name: to_date, dtype: object

Question 9: What is the most recent date in the to_date column?

In [18]:
titles_df['to_date'].sort_values().tail(1) # Most recent to_date below

443307    9999-01-01
Name: to_date, dtype: object

In [19]:
min(titles_df['to_date']) ##some more wizardry

datetime.date(1985, 3, 1)

In [20]:
max(titles_df['to_date']) ##some more wizardry


datetime.date(9999, 1, 1)

In [21]:
pd.read_sql('SELECT COUNT(DISTINCT title) FROM titles', url)

COUNT(DISTINCT title)
0                      7

In [22]:
pd.read_sql("SELECT max(to_date) as most_recent_date FROM titles WHERE to_date NOT LIKE '9999%'", url) ## Using SQL to filter out the current 
## employees that have there to_date's set to 9999-01-01

most_recent_date
0       2002-08-01

## Exercises Part II Advanced Dataframes

In [23]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [24]:
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

Question 2: What is the result of using a right join on the DataFrames?

In [25]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

Question 3: What is the result of using an outer join on the DataFrames?

In [26]:
users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

Question 4: What happens if you drop the foreign keys from the DataFrames and try to merge them?



In [27]:
users.merge(roles, how='outer', indicator=True)

id       name  role_id      _merge
0   1        bob      1.0   left_only
1   2        joe      2.0   left_only
2   3      sally      3.0   left_only
3   4       adam      3.0   left_only
4   5       jane      NaN   left_only
5   6       mike      NaN   left_only
6   1      admin      NaN  right_only
7   2     author      NaN  right_only
8   3   reviewer      NaN  right_only
9   4  commenter      NaN  right_only

# Load the mpg dataset from PyDataset

In [28]:
mpg = data('mpg')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

For Question 7 we can see there are 234 rows and 11 columns after we load the data into our dataframe

In [29]:
## also can use
mpg.shape

(234, 11)

Question 8: Check out your column names and perform any cleanup you may want on them.

In [30]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [31]:
mpg = mpg.rename(columns={'cty': 'city', 'hwy': 'highway', 'displ': 'eng(L)', 'fuelType': 'fuel_type'})
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway fl  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...     ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

Question 9: Display the summary statistics for the dataset

In [32]:
mpg.describe(include = 'all')  #by using the include all in a describe we have it produce the summary statistics
# for all the columns

manufacturer        model      eng(L)         year         cyl  \
count           234          234  234.000000   234.000000  234.000000   
unique           15           38         NaN          NaN         NaN   
top           dodge  caravan 2wd         NaN          NaN         NaN   
freq             37           11         NaN          NaN         NaN   
mean            NaN          NaN    3.471795  2003.500000    5.888889   
std             NaN          NaN    1.291959     4.509646    1.611534   
min             NaN          NaN    1.600000  1999.000000    4.000000   
25%             NaN          NaN    2.400000  1999.000000    4.000000   
50%             NaN          NaN    3.300000  2003.500000    6.000000   
75%             NaN          NaN    4.600000  2008.000000    8.000000   
max             NaN          NaN    7.000000  2008.000000    8.000000   

           trans  drv        city     highway   fl class  
count        234  234  234.000000  234.000000  234   234  
unique        10    3         NaN         NaN    5     7  
top     auto(l4)    f         NaN         NaN    r   suv  
freq          83  106         NaN         NaN  168    62  
mean         NaN  NaN   16.858974   23.440171  NaN   NaN  
std          NaN  NaN    4.255946    5.954643  NaN   NaN  
min          NaN  NaN    9.000000   12.000000  NaN   NaN  
25%          NaN  NaN   14.000000   18.000000  NaN   NaN  
50%          NaN  NaN   17.000000   24.000000  NaN   NaN  
75%          NaN  NaN   19.000000   27.000000  NaN   NaN  
max          NaN  NaN   35.000000   44.000000  NaN   NaN

Question 10: How many different vehicle manufacturers are there?

In [33]:
mpg.manufacturer.value_counts().size 

15

In [34]:
mpg.manufacturer.agg('nunique')

15

In [35]:
mpg.manufacturer.nunique()

15

Question 11: How many different models are there?

In [36]:
mpg.model.value_counts().size

38

In [37]:
mpg.model.agg('nunique')

38

In [38]:
mpg.model.nunique()

38

Questions 12 & 13: Create mileage_difference columns and average_mileage columns for the vehicles

In [39]:
mpg['average_mileage2'] = mpg[['city','highway']].agg('mean', axis = 1)
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway fl  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...     ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26  p   

       class  average_mileage2  
1    compact              23.5  
2    compact              25.0  
3    compact              25.5  
4    compact              25.5  
5    compact              21.0  
..       ...               ...  
230  midsize              23.5  
231  midsize              25.0  
232  midsize              21.0  
233  midsize              22.0  
234  midsize              21.5  

[234 rows x 12 columns]

In [40]:
mpg['mileage_difference'] = mpg.highway - mpg.city
mpg['average_mileage'] = (mpg.city + mpg.highway) / 2
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway fl  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...     ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26  p   

       class  average_mileage2  mileage_difference  average_mileage  
1    compact              23.5                  11             23.5  
2    compact              25.0                   8             25.0  
3    compact              25.5                  11             25.5  
4    compact              25.5                   9             25.5  
5    compact              21.0                  10             21.0  
..       ...               ...                 ...              ...  
230  midsize              23.5                   9             23.5  
231  midsize              25.0                   8             25.0  
232  midsize              21.0                  10             21.0  
233  midsize              22.0                   8             22.0  
234  midsize              21.5                   9             21.5  

[234 rows x 14 columns]

Question 14: Create a column named is_automatic determining if they are automatic or not

In [41]:
mpg['is_automatic'] = np.where(mpg.trans.str.count('auto'), 'automatic', 'manual')
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway fl  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...     ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26  p   

       class  average_mileage2  mileage_difference  average_mileage  \
1    compact              23.5                  11             23.5   
2    compact              25.0                   8             25.0   
3    compact              25.5                  11             25.5   
4    compact              25.5                   9             25.5   
5    compact              21.0                  10             21.0   
..       ...               ...                 ...              ...   
230  midsize              23.5                   9             23.5   
231  midsize              25.0                   8             25.0   
232  midsize              21.0                  10             21.0   
233  midsize              22.0                   8             22.0   
234  midsize              21.5                   9             21.5   

    is_automatic  
1      automatic  
2         manual  
3         manual  
4      automatic  
5      automatic  
..           ...  
230    automatic  
231       manual  
232    automatic  
233       manual  
234    automatic  

[234 rows x 15 columns]

In [42]:
mpg['is_automatic'] = np.where(mpg.trans.str.count('auto'), True, False)
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway fl  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...     ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26  p   

       class  average_mileage2  mileage_difference  average_mileage  \
1    compact              23.5                  11             23.5   
2    compact              25.0                   8             25.0   
3    compact              25.5                  11             25.5   
4    compact              25.5                   9             25.5   
5    compact              21.0                  10             21.0   
..       ...               ...                 ...              ...   
230  midsize              23.5                   9             23.5   
231  midsize              25.0                   8             25.0   
232  midsize              21.0                  10             21.0   
233  midsize              22.0                   8             22.0   
234  midsize              21.5                   9             21.5   

     is_automatic  
1            True  
2           False  
3           False  
4            True  
5            True  
..            ...  
230          True  
231         False  
232          True  
233         False  
234          True  

[234 rows x 15 columns]

Question 15: Using the mpg dataset, find out which manufacturer has the best miles per gallon on average?

In [43]:
mpg.groupby('manufacturer').average_mileage.agg('mean').sort_values(ascending=False)

manufacturer
honda         28.500000
volkswagen    25.074074
hyundai       22.750000
subaru        22.428571
audi          22.027778
toyota        21.720588
pontiac       21.700000
nissan        21.346154
chevrolet     18.447368
ford          16.680000
mercury       15.625000
jeep          15.562500
dodge         15.540541
lincoln       14.166667
land rover    14.000000
Name: average_mileage, dtype: float64

Question 16: Do automatic or manual cars have better miles per gallon?

In [44]:
mpg.groupby('is_automatic').average_mileage.agg('mean').sort_values() #Looks like manual cars are the winners here

is_automatic
True     19.130573
False    22.227273
Name: average_mileage, dtype: float64

## Exercise Part III
-Working with chiptole database

In [57]:
url = get_db_url(host,user,password)
pd.read_sql('SELECT * FROM orders LIMIT 5 OFFSET 50', url) ##using my new input url function to display orders
## from chipotle

Please enter the database you wish to connect to: chipotle


id  order_id  quantity        item_name  \
0  51        23         1    Steak Burrito   
1  52        23         2      Canned Soda   
2  53        24         1  Chicken Burrito   
3  54        24         1      Canned Soda   
4  55        25         1       Steak Bowl   

                                                                          choice_description  \
0   [Roasted Chili Corn Salsa (Medium), [Rice, Fajita Veggies, Cheese, Sour Cream, Lettuce]]   
1                                                                             [Mountain Dew]   
2  [Roasted Chili Corn Salsa (Medium), [Black Beans, Rice, Fajita Veggies, Cheese, Sour C...   
3                                                                                   [Sprite]   
4       [Fresh Tomato Salsa (Mild), [Black Beans, Rice, Fajita Veggies, Cheese, Sour Cream]]   

  item_price  
0     $8.99   
1     $2.18   
2    $10.98   
3     $1.09   
4     $8.99

In [77]:
orders_df2 = pd.read_sql('SELECT * FROM orders', url)
orders_df2

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

Question 2: What is the total price for each order?

In [92]:
orders_df2['item_price'] = orders_df2['item_price'].str.replace('$','').astype(float)
orders_df2

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

      item_price  item.price  
0           2.39        2.39  
1           3.39        3.39  
2           3.39        3.39  
3           2.39        2.39  
4          16.98       16.98  
...          ...         ...  
4617       11.75       11.75  
4618       11.75       11.75  
4619       11.25       11.25  
4620        8.75        8.75  
4621        8.75        8.75  

[4622 rows x 7 columns]

In [94]:
orders_df2.groupby('order_id').sum('item_price')

id  quantity  item_price  item.price
order_id                                         
1            10         4       11.56       11.56
2             5         2       16.98       16.98
3            13         2       12.67       12.67
4            17         2       21.00       21.00
5            21         2       13.70       13.70
...         ...       ...         ...         ...
1830       9223         2       23.00       23.00
1831      13842         3       12.90       12.90
1832       9233         2       13.20       13.20
1833       9237         2       23.50       23.50
1834      13863         3       28.75       28.75

[1834 rows x 4 columns]

Question 3: What are the 3 most popular items?

In [78]:
orders_df2.item_name.value_counts().head(3) ##The three most popular items are listed below using
## value counts and the head(3) 

Chicken Bowl           726
Chicken Burrito        553
Chips and Guacamole    479
Name: item_name, dtype: int64

Question 4: Which item has produced the most revenue?

In [99]:
orders_df2.groupby('item_name').sum('item_price').sort_values(by= 'item_price', ascending = False).head(1)

id  order_id  quantity  item_price  item.price
item_name                                                        
Chicken Bowl  1780635    713926       761     7342.73     7342.73

Question 5: Using the title DataFrame, visualize the number of employees with each title

In [101]:
titles_df

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

In [105]:
employees_df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [104]:
titles_df['title'].value_counts() ##The number of employees with each title

Engineer              115003
Staff                 107391
Senior Engineer        97750
Senior Staff           92853
Technique Leader       15159
Assistant Engineer     15128
Manager                   24
Name: title, dtype: int64

Question 6: Join the employees and titles DataFrames together 

In [111]:
joined_df = titles_df.merge(employees_df, left_on='emp_no', right_on='emp_no', how='inner', indicator=True)
joined_df

emp_no            title   from_date     to_date  birth_date  \
0        10001  Senior Engineer  1986-06-26  9999-01-01  1953-09-02   
1        10002            Staff  1996-08-03  9999-01-01  1964-06-02   
2        10003  Senior Engineer  1995-12-03  9999-01-01  1959-12-03   
3        10004         Engineer  1986-12-01  1995-12-01  1954-05-01   
4        10004  Senior Engineer  1995-12-01  9999-01-01  1954-05-01   
...        ...              ...         ...         ...         ...   
443303  499997         Engineer  1987-08-30  1992-08-29  1961-08-03   
443304  499997  Senior Engineer  1992-08-29  9999-01-01  1961-08-03   
443305  499998     Senior Staff  1998-12-27  9999-01-01  1956-09-05   
443306  499998            Staff  1993-12-27  1998-12-27  1956-09-05   
443307  499999         Engineer  1997-11-30  9999-01-01  1958-05-01   

       first_name last_name gender   hire_date _merge  
0          Georgi   Facello      M  1986-06-26   both  
1         Bezalel    Simmel      F  1985-11-21   both  
2           Parto   Bamford      M  1986-08-28   both  
3       Chirstian   Koblick      M  1986-12-01   both  
4       Chirstian   Koblick      M  1986-12-01   both  
...           ...       ...    ...         ...    ...  
443303    Berhard    Lenart      M  1986-04-21   both  
443304    Berhard    Lenart      M  1986-04-21   both  
443305   Patricia   Breugel      M  1993-10-13   both  
443306   Patricia   Breugel      M  1993-10-13   both  
443307     Sachin   Tsukuda      M  1997-11-30   both  

[443308 rows x 10 columns]

Question 7: Visualize how frequently employees change titles

In [112]:
pd.crosstab(joined_df.emp_no, joined_df.title, margins =True)

title   Assistant Engineer  Engineer  Manager  Senior Engineer  Senior Staff  \
emp_no                                                                         
10001                    0         0        0                1             0   
10002                    0         0        0                0             0   
10003                    0         0        0                1             0   
10004                    0         1        0                1             0   
10005                    0         0        0                0             1   
...                    ...       ...      ...              ...           ...   
499996                   0         1        0                1             0   
499997                   0         1        0                1             0   
499998                   0         0        0                0             1   
499999                   0         1        0                0             0   
All                  15128    115003       24            97750         92853   

title    Staff  Technique Leader     All  
emp_no                                    
10001        0                 0       1  
10002        1                 0       1  
10003        0                 0       1  
10004        0                 0       2  
10005        1                 0       2  
...        ...               ...     ...  
499996       0                 0       2  
499997       0                 0       2  
499998       1                 0       2  
499999       0                 0       1  
All     107391             15159  443308  

[300025 rows x 8 columns]

In [116]:
joined_df.groupby('emp_no').title.count()

emp_no
10001     1
10002     1
10003     1
10004     2
10005     2
         ..
499995    1
499996    2
499997    2
499998    2
499999    1
Name: title, Length: 300024, dtype: int64

Question 8: For each title, find the hire date of the employee that was hired most recently with that title


In [117]:
joined_df.groupby('title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

Question 9: Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [126]:
sql_query = '''
SELECT *
FROM employees as e
JOIN titles as t USING(emp_no)
JOIN dept_emp as de USING(emp_no)
JOIN departments as d USING(dept_no);
'''

the_big_df = pd.read_sql(sql_query, get_db_url(host,user,password))

Please enter the database you wish to connect to: employees


In [127]:
the_big_df

dept_no  emp_no  birth_date first_name last_name gender   hire_date  \
0         d009   10011  1953-11-07       Mary     Sluis      F  1990-01-22   
1         d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
2         d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
3         d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   
4         d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   
...        ...     ...         ...        ...       ...    ...         ...   
489898    d007  499986  1952-07-22     Nathan     Ranta      F  1985-08-11   
489899    d007  499986  1952-07-22     Nathan     Ranta      F  1985-08-11   
489900    d007  499987  1961-09-05      Rimli    Dusink      F  1998-09-20   
489901    d007  499988  1962-09-28   Bangqing   Kleiser      F  1986-06-06   
489902    d007  499988  1962-09-28   Bangqing   Kleiser      F  1986-06-06   

               title   from_date     to_date   from_date     to_date  \
0              Staff  1990-01-22  1996-11-09  1990-01-22  1996-11-09   
1       Senior Staff  1996-09-20  9999-01-01  1989-09-20  9999-01-01   
2              Staff  1989-09-20  1996-09-20  1989-09-20  9999-01-01   
3       Senior Staff  2000-05-04  9999-01-01  1992-05-04  9999-01-01   
4              Staff  1992-05-04  2000-05-04  1992-05-04  9999-01-01   
...              ...         ...         ...         ...         ...   
489898  Senior Staff  1992-08-11  9999-01-01  1985-08-11  9999-01-01   
489899         Staff  1985-08-11  1992-08-11  1985-08-11  9999-01-01   
489900         Staff  1999-12-21  9999-01-01  1999-12-21  9999-01-01   
489901  Senior Staff  1997-07-25  2001-10-09  1988-07-25  2001-10-09   
489902         Staff  1988-07-25  1997-07-25  1988-07-25  2001-10-09   

               dept_name  
0       Customer Service  
1       Customer Service  
2       Customer Service  
3       Customer Service  
4       Customer Service  
...                  ...  
489898             Sales  
489899             Sales  
489900             Sales  
489901             Sales  
489902             Sales  

[489903 rows x 13 columns]

In [128]:
pd.crosstab(the_big_df.title, the_big_df.dept_name)

dept_name           Customer Service  Development  Finance  Human Resources  \
title                                                                         
Assistant Engineer               298         7769        0                0   
Engineer                        2362        58135        0                0   
Manager                            4            2        2                2   
Senior Engineer                 2027        49326        0                0   
Senior Staff                   13925         1247    12139            12274   
Staff                          16150         1424    13929            14342   
Technique Leader                 309         7683        0                0   

dept_name           Marketing  Production  Quality Management  Research  Sales  
title                                                                           
Assistant Engineer          0        6445                1831       378      0  
Engineer                    0       49649               13852      2986      0  
Manager                     2           4                   4         2      2  
Senior Engineer             0       42205               11864      2570      0  
Senior Staff            13940        1270                   0     11637  36191  
Staff                   16196        1478                   0     13495  41808  
Technique Leader            0        6557                1795       393      0